### PR TITLE
Enhancement of ARCHITECTURE.md Documentation

### LINKED ISSUE #578: Enhancement of ARCHITECTURE.md  Documentation
**Description**:
In order to provide detailed information about software components,
Enhancements to architecture documentation are desired to better understand the system's structure and enable informed design decisions.

**Acceptance Criteria**:

- [x] The architecture documentation should cover all major system components, including their interconnections and data flows.
- [x] Detailed descriptions of each component, including their purpose, responsibilities, and dependencies, must be included.
- [x] Updated architectural diagrams, such as flowcharts, sequence diagrams, and component diagrams, should be incorporated into the documentation.
- [x] The documentation should be accessible to all members of the development team and available in an easily accessible repository.

### PR DESCRIPTION
**Pull request by VictorCavichioli**
Closes #578

### CODE REVIEW

File: docs/ARCHITECTURE.md [hunk 1]
```diff
@@ -9,10 +23,34 @@ can be configured to run only during certain time periods or not at all. Setting
...
+    <br/>
+    <figcaption >Figure 1: ecChronos and Cassandra Nodes.</figcaption>
+  </div>
+</figure>
+
 ## Concepts
 
 ### Leases
 
+In the context of Apache Cassandra, a "lease" refers to a mechanism employed during the repair process within Cassandra.
```
Comment (masokol) [inline docs/ARCHITECTURE.md:38]: I find this information a bit too generic, this document should be in the context of ecChronos not Cassandra. We already explain how ecChronos achieves coordination on the locks(leases) below. Maybe just adding a reference would be enough.

File: docs/ARCHITECTURE.md [hunk 2]
```diff
@@ -60,7 +105,15 @@ For more information about time based run policy refer to [Time based run policy
 The repair scheduling begins by providing a [RepairConfiguration](../core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/repair/RepairConfiguration.java) to the [RepairScheduler](../core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/repair/RepairSchedulerImpl.java).
 The repair scheduler then creates a [TableRepairJob](../core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/repair/TableRepairJob.java)
 or [IncrementalRepairJob](../core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/repair/IncrementalRepairJob.java)
-and schedules it using the [ScheduleManager](../core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/scheduling/ScheduleManagerImpl.java).
+and schedules it using the [ScheduleManager](../core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/scheduling/ScheduleManagerImpl.java) [\[2\]](#references).
+
+<figure>
```
Comment (masokol) [inline docs/ARCHITECTURE.md:110]: This picture is incorrect, it should be
schedule manager -> refresh priorities -> pick job with highest priority -> should job run (run policy)
yes -> Create job tasks -> execute tasks -> schedule manager
no -> take next job -> should job run (run policy)

File: docs/ARCHITECTURE.md [hunk 3]
```diff
@@ -112,7 +165,7 @@ Assuming that X is more than one hour before Y this will produce three sub range
 
 * (0, 15] repaired at X.
 * (15, 20] repaired at Y
-* (20, 30] repaired at X
+* (20, 30] repaired at X5f
```
Comment (masokol) [inline docs/ARCHITECTURE.md:168]: X5f?

**Review by masokol — CHANGES_REQUESTED** [review]

File: docs/images/ecchronos-instances.png
Comment (jwaeab) [inline docs/images/ecchronos-instances.png:1]: Ecchronos -> ecChronos

Also, if the ecChronos instance and Cassandra node runs on the same entity in this picture, perhaps they should be grouped together. Looking at the picture now, it seems all instances/nodes run in seperate environments.

File: docs/ARCHITECTURE.md [hunk 4]
```diff
@@ -9,10 +23,34 @@ can be configured to run only during certain time periods or not at all. Setting
...
+  </div>
+</figure>
+
 ## Concepts
 
 ### Leases
 
+In the context of Apache Cassandra, a "lease" refers to a mechanism employed during the repair process within Cassandra.
+
+Upon the initiation of a repair in a Cassandra cluster, a "lease" is granted to one of the cluster nodes, which serves as the repair coordinator. This coordinating node is responsible for overseeing the repair process to ensure that all data replicas conform and any disparities are addressed.
```
Comment (jwaeab) [inline docs/ARCHITECTURE.md:40]: Why the use of "lease" everywhere? Why not just use lease (without citation) if that's the correct word?

File: docs/ARCHITECTURE.md [hunk 5]
```diff
@@ -9,10 +23,34 @@ can be configured to run only during certain time periods or not at all. Setting
...
+
+The "lease" bestows upon the coordinating node the exclusive right to conduct the repair during a specific time frame. Within this duration, the coordinating node assesses the data, ensuring that all replicas are updated and consistent. Additionally, it helps prevent multiple nodes from concurrently initiating repairs on the same data, thereby mitigating potential consistency issues and cluster overload.
+
+Once the repair is completed by the coordinating node, the "lease" is released, enabling other nodes to request the "lease" and carry out their own repairs as needed. This helps in efficiently distributing the repair load within the cluster [\[1\]](#references).
+
+<figure>
+  <div align="center">
+    <img src="images/election.png" alt="Alt Text" style="display: block; margin-left: auto; margin-right: auto; width: 500px;">
+    <br/>
+    <figcaption >Figure 2: Lease Typically Election.</figcaption>
```
Comment (jwaeab) [inline docs/ARCHITECTURE.md:50]: Unnecessary space after each figcaption tag.

File: docs/images/repair-job-flow.png
Comment (jwaeab) [inline docs/images/repair-job-flow.png:1]: What's up with the black dot? Void? ;-)

What will make the dot do the transition from its state to Schedule Manager again? Why is it marked red?

File: docs/ARCHITECTURE.md (hunk 3 above)
Comment (VictorCavichioli) [reply docs/ARCHITECTURE.md:168]: It was a mistake, sorry.

File: docs/images/repair-job-flow.png
Comment (VictorCavichioli) [reply docs/images/repair-job-flow.png:1]: In states machines it means "end", but based on previously conversations we've decided to change a little bit this diagram and remove this.

File: docs/ARCHITECTURE.md [hunk 6]
```diff
@@ -9,10 +22,50 @@ can be configured to run only during certain time periods or not at all. Setting
 make sure repair is spread out over the interval time while alarms are provided to signal when a job has not run for
 longer than expected.
 
+<div align="center">
```
Comment (masokol) [inline docs/ARCHITECTURE.md:25]: This chart looks really strange, see if you can get it more spaced out. Right now it looks like each ecchronos is connected to 2 cassandra nodes.

File: docs/ARCHITECTURE.md [hunk 7]
```diff
@@ -9,10 +22,50 @@ can be configured to run only during certain time periods or not at all. Setting
...
+
+Once the repair is completed by the coordinating node, the "lease" is released, enabling other nodes to request and carry out their own repairs as needed. It helps in efficiently distributing the repair load within the cluster [\[1\]](#references).
+
+<div align="center">
+
+  ```mermaid
+  flowchart TB
+    A[Create Lease] -->|Failed| b[Sleap]
+    b[Sleap] --> A[Create Lease]
+    A[Create Lease] -->|Succeeded| c[Become Master]
```
Comment (masokol) [inline docs/ARCHITECTURE.md:60]: There's no master concept in ecchronos, you either get the lock or not.

File: docs/ARCHITECTURE.md [hunk 8]
```diff
@@ -60,7 +126,38 @@ For more information about time based run policy refer to [Time based run policy
...
+        direction LR
+        RepairConfiguration1
+        RepairConfiguration2
+        RepairConfiguration3
+      }
+      state SchedulerManager {
+        [...]
+      }
+      state ShouldJobRunCondition <<choice>>
+      RepairScheduler --> CreateRepairJob
```
Comment (masokol) [inline docs/ARCHITECTURE.md:146]: This should be "CreateSchedules" or "CreateRepairJobs"

File: docs/ARCHITECTURE.md [hunk 9]
```diff
@@ -112,7 +209,7 @@ Assuming that X is more than one hour before Y this will produce three sub range
 
 * (0, 15] repaired at X.
 * (15, 20] repaired at Y
-* (20, 30] repaired at X
+* (20, 30] repaired at X5f
```
Comment (masokol) [inline docs/ARCHITECTURE.md:212]: it's still X5f

File: docs/ARCHITECTURE.md [hunk 10]
```diff
@@ -9,10 +22,50 @@ can be configured to run only during certain time periods or not at all. Setting
...
+  ```
+
+  <figcaption>Figure 1: ecChronos and Cassandra Nodes.</figcaption>
+</div>
+
 ## Concepts
 
 ### Leases
 
+In the context of Apache Cassandra, a "lease" refers to a mechanism employed during the repair process within Cassandra.
```
Comment (masokol) [inline docs/ARCHITECTURE.md:46]: This is not true, in context of Cassandra there's no "lease" during repair process. "Lease" is something you have to implement on the application level (like ecchronos, already described below)

File: docs/ARCHITECTURE.md [hunk 11]
```diff
@@ -9,10 +22,50 @@ can be configured to run only during certain time periods or not at all. Setting
...
+
+Upon the initiation of a repair in a Cassandra cluster, it's granted to one of the cluster nodes, which serves as the repair coordinator. This coordinating node is responsible for overseeing the repair process to ensure that all data replicas conform and any disparities are addressed.
```
Comment (masokol) [inline docs/ARCHITECTURE.md:48]: This is not true either, there are no leases in Cassandra.

File: docs/ARCHITECTURE.md [hunk 12]
```diff
@@ -60,7 +126,38 @@ For more information about time based run policy refer to [Time based run policy
...
+      }
+      state ShouldJobRunCondition <<choice>>
+      RepairScheduler --> CreateRepairJob
+      CreateRepairJob --> SchedulerManager
+      SchedulerManager --> RefreshPriorities
+      RefreshPriorities --> PickJobWithHighestPriority
+      PickJobWithHighestPriority --> ShouldJobRun
+      ShouldJobRun --> ShouldJobRunCondition
+      ShouldJobRunCondition --> PickJobWithHighestPriority: No
+      ShouldJobRunCondition --> CreateJobTask: Yes
```
Comment (masokol) [inline docs/ARCHITECTURE.md:153]: CreateJobTasks

**Review by masokol — CHANGES_REQUESTED** [review]

**Review by jwaeab — APPROVED** [review]

**Pull request merged**